### PR TITLE
Fix regression in pip parse for finding implicit namespace packages.

### LINKED ISSUE #499: pip_parse regression in 0.3.0
# 🐞 bug report

### Affected Rule

 The issue is caused by the rule: <code>pip_parse</code>

### Is this a regression? 

Yes, the previous version in which this bug was not present was: 0.2.0

### Description

Depending on google-api-core fails when using pip_parse from 0.3.0, but succeeds in 0.2.0.
pip_install works in both versions.

Originally reported in #431, but was triaged as a new issue here - https://github.com/bazelbuild/rules_python/issues/431#issuecomment-871375398

## 🔬 Minimal Reproduction

https://github.com/caseyduquettesc/rules_python_demo

```
% bazel run //scripts/demo
INFO: Invocation ID: 3dd0bf13-d654-44ff-b250-34e69abb2cd7
INFO: Analyzed target //scripts/demo:demo (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //scripts/demo:demo up-to-date:
  bazel-bin/scripts/demo/demo
INFO: Elapsed time: 0.217s, Critical Path: 0.02s
INFO: 1 process: 1 internal.
INFO: Running command line: bazel-bin/scripts/demo/demo
INFO: Build Event Protocol files produced successfully.
INFO: Build completed successfully, 1 total action
Traceback (most recent call last):
  File "/Users/.cache/_bazel_root/3e48c9752578ebae810f99b25573480c/execroot/demo/bazel-out/darwin-fastbuild/bin/scripts/demo/demo.runfiles/demo/scripts/demo/demo.py", line 1, in <module>
    from google.api_core.exceptions import ResourceExhausted
ModuleNotFoundError: No module named 'google.api_core'
```

## 🔥 Exception or Error

<pre><code>Traceback (most recent call last):
  File "/Users/.cache/_bazel_root/3e48c9752578ebae810f99b25573480c/execroot/demo/bazel-out/darwin-fastbuild/bin/scripts/demo/demo.runfiles/demo/scripts/demo/demo.py", line 1, in <module>
    from google.api_core.exceptions import ResourceExhausted
ModuleNotFoundError: No module named 'google.api_core'
</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre><code>MacOS 10.15.7</code></pre>

**Output of `bazel version`:**

<pre><code>Build label: 4.1.0.188
Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Jun 21 23:14:47 2021 (1624317287)
Build timestamp: 1624317287
Build timestamp as int: 1624317287</code></pre>

**Rules_python version:**

<pre><code>0.3.0</code></pre>

**Anything else relevant?**
Originally posted in https://github.com/bazelbuild/rules_python/issues/431#issuecomment-868933316

### PR DESCRIPTION
**Pull request by hrfuller**
fixes #499 
Pathlib.Path normalizes path names, which caused lookups of relative
paths using the special '.' directory path to fail to find parents of
standard packages.

Drive-by change fixes passing quiet argument to child repos.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Does not include precompiled binaries, eg. `.par` files. See [CONTRIBUTING.md](https://github.com/bazelbuild/rules_python/blob/master/CONTRIBUTING.md) for info
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
pip_parse fails to find implicit namespace packages.

Issue Number: #499 

## What is the new behavior?
pip_parse can find implicit namespace packages.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by thundergolfer** [comment]
Using the https://github.com/caseyduquettesc/rules_python_demo repro I still have the `ModuleNotFoundError: No module named 'google.api_core'` error. 

Running `PYTHONVERBOSE=2 bazel run //scripts/...` I can see that Python isn't checking in `deps_pypi__google_api_core/google/` for the `api_core` module.

**Comment by hrfuller** [comment]
Mine definitely is
```
bytecode is stale for 'google.api_core'
# code object from /private/var/tmp/_bazel_hfuller/67a2eb7ad29da3877a82ffb129fac926/execroot/demo/bazel-out/darwin-fastbuild/bin/scripts/demo/demo.runfiles/deps_pypi__google_api_core/google/api_core/__init__.
py
```
What interpreter version do you have on the path?

**Comment by hrfuller** [comment]
Hey @caseyduquettesc can you test out this branch against your demo repository and see if it fixes the issue you were having?

**Comment by caseyduquettesc** [comment]
> can you test out this branch against your demo repository and see if it fixes the issue you were having?

Sure thing. Will get back to you in a couple hours, just got back from some time off.

**Comment by caseyduquettesc** [comment]
@hrfuller I used your branch in the example at https://github.com/caseyduquettesc/rules_python_demo and in a private project that has its own python toolchain (3.8.10) and was unable to reproduce the issue in both. Functionally, the change LGTM.

**Comment by ssiano** [comment]
@hrfuller, I verified that this PR also fixes the [ModuleNotFoundError: No module named 'azure.storage'](https://github.com/bazelbuild/rules_python/pull/483#issuecomment-872682591).

I was using Python 3.6.

Cc: @thundergolfer

**Comment by thundergolfer** [comment]
Can also confirm this removes the `ModuleNotFound` errors in the repro I'm using. Tested in https://github.com/thundergolfer-forks/rules_python_demo 

```
from google.api_core.exceptions import ResourceExhausted
import google.api_core
import google.auth
```

All that succeeds.

File: python/pip_install/extract_wheels/lib/namespace_pkgs.py [hunk 1]
```diff
@@ -47,7 +50,7 @@ def implicit_namespace_packages(
     return namespace_pkg_dirs
```
Comment (thundergolfer) [inline python/pip_install/extract_wheels/lib/namespace_pkgs.py:50]: Function now returning `Set[pathlib.Path]` not `Set[str]`

File: python/pip_install/extract_wheels/lib/namespace_pkgs.py [hunk 2]
```diff
@@ -23,19 +23,22 @@ def implicit_namespace_packages(
     namespace_pkg_dirs: Set[str] = set()
     standard_pkg_dirs: Set[str] = set()
     # Traverse bottom-up because a directory can be a namespace pkg because its child contains module files.
+    directory = pathlib.Path(directory)
+    ignored_dirnames = [pathlib.Path(p) for p in ignored_dirnames or ()]
     for dirpath, dirnames, filenames in os.walk(directory, topdown=False):
+        dirpath = pathlib.Path(dirpath)
```
Comment (thundergolfer) [inline python/pip_install/extract_wheels/lib/namespace_pkgs.py:29]: nit: easier to type-check with `mypy` if variable names don't  get reassigned with a type change. Could instead be `directory_path` (L26) and `for dirpath_str, dirnames, filenames`.

**Review by thundergolfer — APPROVED** [review]
Needs a fix to types, but otherwise good to merge 👍

**Pull request merged**